signac/core/collection_api.py

```diff
@@ -0,0 +1,183 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from contextlib import contextmanager
+from collections import defaultdict
+from abc import abstractmethod
+from abc import ABCMeta
+
+try:
+    import numpy
+    NUMPY = True
+except ImportError:
+    NUMPY = False
+
+try:
+    from collections.abc import Collection
+except ImportError:
+    # Collection does not exist in Python 3.5, only Python 3.6 or newer.
+
+    from collections.abc import Sized, Iterable, Container
+
+    def _check_methods(C, *methods):
+        mro = C.__mro__
+        for method in methods:
+            for B in mro:
+                if method in B.__dict__:
+                    if B.__dict__[method] is None:
+                        return NotImplemented
+                    break
+            else:
+                return NotImplemented
+        return True
+
+    class Collection(Sized, Iterable, Container):  # type: ignore
+        @classmethod
+        def __subclasshook__(cls, C):
+            if cls is Collection:
+                return _check_methods(C,  "__len__", "__iter__", "__contains__")
+            return NotImplemented
+
+
+class CustomABCMeta(ABCMeta):
+    """ Metaclass for the definition of SyncedCollection.
+
+    This metaclass automatically registers Synced Classes definitions,
+    which enables the automatic determination of Synced Class for a
+    base type and backend.
+    """
+    def __init__(cls, name, bases, dct):
+        if not hasattr(cls, 'registry'):
+            cls.registry = defaultdict(list)
+        else:
+            if not bool(cls.__abstractmethods__):
+                cls.registry[cls.backend].append(cls)
+        return super().__init__(name, bases, dct)
+
+
+class SyncedCollection(Collection, metaclass=CustomABCMeta):
+    """The base synced collection represents a collection that is synced with a
+    file.
+    The class is intended for use as an ABC.In addition, it declares abstract
+    methods that must be implemented by any subclass.
+    """
+    base_type = None
+    backend = None
+
+    def __init__(self, parent=None):
+        self._data = None
+        self._parent = parent
+        self._suspend_sync_ = 0
+
+    @classmethod
+    def from_base(cls, data, backend=None, **kwargs):
+        """This method dynamically resolve the type of object to the
+        corresponding synced collection.
+
+        Parameters
+        ----------
+        data : any
+            Data to be converted from base class.
+        filename: str
+            Name of file to store the data (Default value = None).
```

Comment:
I think we still want to assume that every synced collection is synchronized with _something_, right? It might be a Redis cache or a MySQL db, but we still assume synchronization is happening? The question of how to handle a single centralized database as opposed to our current distributed model is something I think we won't be able to address until we try such a back end (which is why we *must* implement additional backends before we can consider this branch feature complete), but I suspect you'd still want to store the filename (or resource) in every collection, except that multiple collections could point to the same place. The `sync` method would have to appropriately handle this for a particular backend.

If you're asking about the naming, we can certainly choose something other than "file".

Reply:
Well, if it is related to some abstract resource then we **have** to call it something other than file.

Reply:
Is the assumption that *any* resource that our backends supports can be described by a single string argument? Like a URI? That is a potential design choice, but we need to be intentional about it.

Reply:
https://github.com/glotzerlab/signac/pull/336

Reply:
This is a mistake. `Filename` is intended to be a backend-specific. I removed the `filename` from `SyncedCollection` but forgot the documentation. I'll correct it.